File: bindings/javascript/src/safe_bloom.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeBloom - Probabilistic set membership (Bloom filter).
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 * @module
 */

import { getLib } from './ffi.js';
import { ok, err } from './result.js';

/**
 * Bloom filter backed by the libproven FFI.
 */
export class BloomFilter {
  /** @type {Deno.PointerObject|null} */
  #ptr = null;

  /**
   * Create a bloom filter.
   *
   * @param {number} expectedElements - Expected number of elements.
   * @param {number} falsePositiveRate - Desired false positive rate (0-1).
   * @returns {{ ok: true, value: BloomFilter } | { ok: false, error: string }}
   */
  static create(expectedElements, falsePositiveRate) {
    const symbols = getLib();
    const ptr = symbols.proven_bloom_create(expectedElements, falsePositiveRate);
    if (ptr === null) return err('Failed to create bloom filter');
    const filter = new BloomFilter();
    filter.#ptr = ptr;
    return ok(filter);
  }

  /**
   * Add an element to the filter.
   *
   * @param {Uint8Array|string} element - The element to add.
   */
  add(element) {
    if (this.#ptr === null) return;
    const symbols = getLib();
    const bytes = typeof element === 'string' ? new TextEncoder().encode(element) : element;
    symbols.proven_bloom_add(this.#ptr, bytes, bytes.length);
  }

  /**
   * Check if an element might be in the filter.
   *
   * @param {Uint8Array|string} element - The element to check.
   * @returns {boolean} True if the element might be in the filter (may be false positive).
   */
  contains(element) {
    if (this.#ptr === null) return false;
    const symbols = getLib();
    const bytes = typeof element === 'string' ? new TextEncoder().encode(element) : element;
    return symbols.proven_bloom_contains(this.#ptr, bytes, bytes.length);
  }

  /**
   * Close and free the native bloom filter.
   */
  close() {
    if (this.#ptr !== null) {
      const symbols = getLib();
      symbols.proven_bloom_free(this.#ptr);
      this.#ptr = null;
    }
  }
}

File: bindings/python/proven/safe_bloom.py
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

"""
SafeBloom - Probabilistic set membership with bloom filters.

Provides space-efficient probabilistic set membership testing.
All operations are delegated to the Idris core via FFI using opaque handles.
"""

from typing import Union

from .core import ProvenStatus, get_lib


class BloomFilter:
    """
    Bloom filter for probabilistic set membership via FFI.

    A bloom filter can tell you:
    - Definitely NOT in the set (no false negatives)
    - POSSIBLY in the set (may have false positives)

    Example:
        >>> bf = BloomFilter(1000, 0.01)  # 1000 items, 1% false positive rate
        >>> bf.insert("hello")
        >>> bf.contains("hello")
        True
        >>> bf.contains("world")  # Probably False
        False
    """

    def __init__(self, expected_items: int, false_positive_rate: float = 0.01):
        """
        Create a bloom filter via FFI.

        Args:
            expected_items: Expected number of items to insert
            false_positive_rate: Desired false positive rate (0-1)

        Raises:
            ValueError: If parameters are invalid
        """
        if expected_items <= 0:
            raise ValueError("Expected items must be positive")
        if not 0 < false_positive_rate < 1:
            raise ValueError("False positive rate must be between 0 and 1")

        lib = get_lib()
        result = lib.proven_bloom_create(expected_items, false_positive_rate)
        if result.status != ProvenStatus.OK or result.handle is None:
            raise RuntimeError("Failed to create bloom filter via FFI")

        self._handle = result.handle
        self._expected_items = expected_items
        self._false_positive_rate = false_positive_rate
        self._count = 0
        self._lib = lib

    def __del__(self):
        """Free the FFI bloom filter handle."""
        if hasattr(self, "_handle") and self._handle is not None:
            try:
                self._lib.proven_bloom_free(self._handle)
            except Exception:
                pass

    @property
    def count(self) -> int:
        """Get number of items inserted."""
        return self._count

    def insert(self, item: Union[str, bytes]) -> None:
        """
        Insert an item into the filter via FFI.

        Args:
            item: Item to insert
        """
        if isinstance(item, str):
            item = item.encode("utf-8")
        status = self._lib.proven_bloom_insert(self._handle, item, len(item))
        if status == ProvenStatus.OK:
            self._count += 1

    def contains(self, item: Union[str, bytes]) -> bool:
        """
        Check if item might be in the filter via FFI.

        Args:
            item: Item to check

        Returns:
            True if possibly in set, False if definitely not
        """
        if isinstance(item, str):
            item = item.encode("utf-8")
        result = self._lib.proven_bloom_contains(self._handle, item, len(item))
        if result.status != ProvenStatus.OK:
            return False
        return result.value

    def __contains__(self, item: Union[str, bytes]) -> bool:
        """Support 'in' operator."""
        return self.contains(item)

    def clear(self) -> None:
        """Reset the filter by recreating via FFI."""
        self._lib.proven_bloom_free(self._handle)
        result = self._lib.proven_bloom_create(self._expected_items, self._false_positive_rate)
        if result.status == ProvenStatus.OK and result.handle is not None:
            self._handle = result.handle
        self._count = 0
